### PR TITLE
feat(python-sdk): support streaming file uploads

### LINKED ISSUE #1376: Python client: Filesystem.write only accepts bytes/str, forcing whole payloads into memory — accept file-like objects and stream
### What would you like to be added

`Filesystem.write` / `AsyncFilesystem.write` (k8s-agent-sandbox 0.5.4) accept `content: bytes | str`, so a caller must materialize the entire payload in memory before uploading. For large workspace files (datasets, archives, model outputs) that is a transient memory spike on the client, and callers integrating the sandbox behind streaming-oriented interfaces (e.g. the OpenAI Agents SDK's sandbox session contract, which hands backends an `io.IOBase` precisely so they can stream) end up buffering or imposing artificial size caps.

The wall is thin, because both ends already stream:

- the transport (`httpx` / `requests`) natively streams file-like objects in multipart payloads — `files={"file": (path, fileobj)}` works today;
- the runtime server receives the upload through FastAPI's `UploadFile`, which spools to disk rather than holding memory.

Only the client signature forces the buffering.

### Proposed change

Widen the accepted type and pass file-like objects through:

```python
async def write(self, path: str, content: bytes | str | BinaryIO, ...):
    if isinstance(content, str):
        content = content.encode("utf-8")
    # bytes -> as today; file-like -> handed to the transport, which streams it
    files_payload = {"file": (path, content)}
```

The only wrinkle is the tracing attribute `sandbox.file.size`, which calls `len(content)` — for file-like objects it could be omitted or derived via `os.fstat` when available.

Happy to send this as a PR (CLA in progress on our side); filing the issue first per the contributor guide.

/kind feature

### PR DESCRIPTION
**Pull request by HasonoCell**
#### What this PR does / why we need it:

Python SDK callers currently have to materialize large uploads as `bytes` because `Filesystem.write` and `AsyncFilesystem.write` do not accept file objects. This change lets both methods accept binary file objects and stream them from their current position in bounded chunks.

- streams legacy uploads with a single-file multipart body without buffering the payload
- streams sandboxd raw PUT uploads and disables replay for non-rewindable request bodies
- preserves the existing `bytes` and `str` behavior and omits unknown stream sizes from tracing
- documents stream ownership and current-position semantics, with unit and E2E coverage

Validation: the Python 3.11 unit suite passes with 381 passed, 1 skipped, and 8 subtests passed. A local HTTP smoke test also verified chunked sync and async multipart uploads.

#### Which issue(s) this PR is related to:

Fixes #1376

#### Release Note

```release-note
Python SDK file writes now accept binary file objects and stream uploads without buffering the entire payload in client memory.
```

## Summary by CodeRabbit

* **New Features**
  * File uploads now accept binary file objects alongside strings and bytes.
  * Streams upload from their current position in bounded chunks without loading the entire file into memory.
  * Streaming uploads support synchronous and asynchronous workflows, including seekable and non-seekable streams.
  * Source streams remain open after uploading.
  * Streaming requests are sent only once to prevent replaying file content.

* **Documentation**
  * Added guidance on file ownership, single-use uploads, and possible position changes after failures.

### CODE REVIEW

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/filesystem.py [hunk 1]
```diff
@@ -12,17 +12,75 @@
...
+
+def _multipart_framing(path: str) -> tuple[bytes, bytes, str]:
+    """Build the framing for a streaming, single-file multipart request."""
+    boundary = secrets.token_hex(16)
+    field = RequestField(name="file", data=None, filename=path)
+    field.make_multipart(content_disposition="form-data")
+    prefix = (
+        f"--{boundary}\r\n".encode()
+        + field.render_headers().encode("utf-8")
+    )
```
[the remark this answers is not in the document]
Comment (HasonoCell) [reply clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/filesystem.py:68]: `RequestField.render_headers()` already terminates the header block with `\r\n\r\n`; this behavior is consistent in urllib3 1.26, 2.0, and 2.7. The existing multipart tests also pass (`86 passed, 1 skipped`). Adding another CRLF here would instead prepend an extra newline to the uploaded file payload, so no code change is needed.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/connector.py [hunk 2]
```diff
@@ -605,7 +608,8 @@ def send_request(self, method: str, endpoint: str, **kwargs) -> requests.Respons
             kwargs.pop("allow_redirects", None)
 
             # Send the request with redirections blocked
-            response = self.session.request(method, url, allow_redirects=False, **kwargs)
+            request = requests.request if disable_retries else self.session.request
```
Comment (aditya-shantanu) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/connector.py:628]: Worth a brief comment on why the session is bypassed: the mounted Retry adapter would replay a consumed generator as an empty body. Note this also drops the body-safe connect retries, so streamed PUTs fail on a single connect hiccup.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/filesystem.py [hunk 3]
```diff
@@ -66,14 +130,26 @@ def write(
...
+            data = _iter_binary_chunks(content) if streaming else content
             self.connector.send_request(
                 "PUT", _sandboxd_files_endpoint(path),
-                data=content,
+                data=data,
                 headers={"Content-Type": "application/octet-stream"},
                 timeout=timeout,
+                _disable_retries=streaming,
+            )
+        elif streaming:
```
Comment (aditya-shantanu) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/filesystem.py]: This path is replay-safe only because POST is not in RETRYABLE_METHODS; if POST ever became retryable, urllib3 would resend the exhausted generator as an empty body (silent truncation). A short comment recording that invariant here and in async_filesystem would guard against that.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/filesystem.py [hunk 4]
```diff
@@ -12,17 +12,75 @@
...
 import logging
 import os
 import posixpath
+import secrets
 import urllib.parse
-from typing import List
+from collections.abc import Iterator
+from typing import BinaryIO, List
+
+from urllib3.fields import RequestField
```
Comment (aditya-shantanu) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/filesystem.py:26]: This makes urllib3 a direct import while it is only a transitive dep via requests; consider declaring it in pyproject.toml dependencies.

**Review by aditya-shantanu — COMMENTED** [review]
Nice work — streaming semantics, replay safety, and sync/async parity are all handled carefully, and the test coverage is thorough. Three non-blocking suggestions inline.

**Comment by aditya-shantanu** [comment]
Re-reviewed at a1c91c8. Delta from ee5ddbd is the rebase plus my three notes addressed: retry-bypass rationale documented in connector.py, POST-not-in-RETRYABLE_METHODS invariant recorded in filesystem.py/async_filesystem.py, and urllib3 declared in pyproject.toml. Streaming/replay behavior unchanged.

/lgtm

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/connector.py [hunk 5]
```diff
@@ -616,8 +620,13 @@ def send_request(self, method: str, endpoint: str, **kwargs : Any) -> requests.R
...
             kwargs.pop("allow_redirects", None)
 
-            # Send the request with redirections blocked
-            response = self.session.request(method, url, allow_redirects=False, **kwargs)
+            # A streamed body cannot be replayed by the Session's Retry
+            # adapter: a retry would send the consumed generator as an empty
+            # body. Bypassing the Session also means that a streamed PUT gets
+            # only one connection attempt, so callers see transient connect
+            # failures directly.
+            request = requests.request if disable_retries else self.session.request
```
Comment (janetkuo) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/connector.py]: Bypassing `self.session` prevents replaying the consumed generator on 5xx retries, though it also creates an ephemeral session that bypasses connection pooling and any custom transport adapters mounted on `self.session`. For large streaming uploads where avoiding replay corruption is essential, this tradeoff makes sense.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/filesystem.py [hunk 6]
```diff
@@ -12,17 +12,75 @@
...
+
+
+def _multipart_framing(path: str) -> tuple[bytes, bytes, str]:
+    """Build the framing for a streaming, single-file multipart request."""
+    boundary = secrets.token_hex(16)
+    field = RequestField(name="file", data=b"", filename=path)
+    field.make_multipart(content_disposition="form-data")
+    prefix = (
+        f"--{boundary}\r\n".encode()
+        + field.render_headers().encode("utf-8")
```
Comment (janetkuo) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/filesystem.py]: nit: Encoding calls are slightly inconsistent (`f"--{boundary}\r\n".encode()` vs `field.render_headers().encode("utf-8")`). Both default to UTF-8 in Python 3, but using `.encode("utf-8")` consistently reads cleaner.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/async_filesystem.py [hunk 7]
```diff
@@ -12,16 +12,36 @@
...
 from k8s_agent_sandbox.trace_manager import async_trace_span, trace
 
 
+async def _iter_multipart_body(
+    prefix: bytes,
+    content: BinaryIO,
+    suffix: bytes,
+) -> AsyncIterator[bytes]:
+    yield prefix
+    while chunk := await asyncio.to_thread(_read_binary_chunk, content):
```
Comment (janetkuo) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/async_filesystem.py]: Offloading `_read_binary_chunk` via `asyncio.to_thread` ensures synchronous disk I/O does not block the event loop while `httpx` streams chunks out.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/test/unit/test_filesystem.py [hunk 8]
```diff
@@ -155,5 +196,164 @@ def _do_write(self, **kwargs):
...
+    def test_string_tracing_uses_utf8_byte_size(self, get_current_span):
+        span = get_current_span.return_value
+        span.is_recording.return_value = True
+
+        self.filesystem.write("unicode.txt", "你好")
+
+        span.set_attribute.assert_any_call("sandbox.file.size", 6)
+
+
+class TestAsyncFilesystemStreamingWrite(unittest.IsolatedAsyncioTestCase):
```
Comment (janetkuo) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/test/unit/test_filesystem.py]: Consider adding async test cases for non-seekable streams, empty files, and path validation rejection before stream reads to mirror `TestFilesystemStreamingWrite` and guarantee full sync/async test parity.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/connector.py (hunk 5 above)
Comment (janetkuo) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/connector.py]: Bypassing `self.session` with `requests.request` strips connection pooling, persistent TCP/keep-alive, configured proxy settings, SSL verification certificates, and session-level headers. Each streaming upload ends up paying the full handshake cost and tearing down the socket.

In `async_connector.py`, retries are disabled per-request while preserving `self.client` by dynamically setting `max_retries = 0 if disable_retries else MAX_RETRIES`. For `requests.Session`, we should keep requests on the session—either by mounting a separate non-retrying adapter, preparing the request and sending with a zero-retry adapter, or configuring the retry adapter so the session and connection pool are preserved.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/async_filesystem.py [hunk 9]
```diff
@@ -12,19 +12,44 @@
...
+        yield chunk
+    yield suffix
+
+
+async def _iter_binary_chunks(content: BinaryIO) -> AsyncIterator[bytes]:
+    """Read a binary stream in bounded chunks without blocking the loop."""
+    while chunk := await asyncio.to_thread(_read_binary_chunk, content):
```
Comment (janetkuo) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/async_filesystem.py]: Scheduling every 64 KiB chunk via `asyncio.to_thread(_read_binary_chunk, content)` means a 100 MB upload dispatches ~1,600 tasks to the default `ThreadPoolExecutor`. For streaming file uploads, this creates noticeable context-switching overhead across the event loop and thread pool.

Consider increasing `_STREAM_CHUNK_SIZE` (e.g. 256 KiB–1 MiB) or exposing buffer size as an option to reduce thread hops for large files.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/async_filesystem.py [hunk 10]
```diff
@@ -38,18 +63,22 @@ def __init__(
     @async_trace_span("write")
     async def write(
         self,
-        path: str, content: bytes | str,
+        path: str,
+        content: bytes | str | BinaryIO,
```
Comment (janetkuo) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/async_filesystem.py]: `content` accepts `BinaryIO`, which implies synchronous readers like `open(..., "rb")` or `BytesIO`. However, users working in `AsyncFilesystem` may attempt to pass an async file-like object (such as `aiofiles` or `anyio` file streams) because they have a `.read()` method. When passed here, `content.read()` returns a coroutine rather than bytes, failing downstream in `_read_binary_chunk` with a `TypeError`.

We should clarify in the docstring and type hints that only synchronous binary file streams are supported here, or note if async stream support is planned separately.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/filesystem.py [hunk 11]
```diff
@@ -12,19 +12,74 @@
...
+        raise TypeError("Binary file objects must return bytes from read().")
+    return bytes(chunk)
+
+
+def _iter_binary_chunks(content: BinaryIO) -> Iterator[bytes]:
+    while chunk := _read_binary_chunk(content):
+        yield chunk
+
+
+def _multipart_framing(path: str) -> tuple[bytes, bytes, str]:
```
Comment (janetkuo) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/filesystem.py]: `_multipart_framing` cleanly constructs the multipart boundary and header framing using `urllib3.fields.RequestField` without buffering the stream body in memory. Nice handling of chunked transfer encoding for the legacy runtime.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/connector.py (hunk 5 above)
Comment (barney-s) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/connector.py:628]: Bypassing the connection pool of `self.session` with `requests.request` means a new TCP connection (and potentially a TLS handshake) is opened and closed for every single streamed file upload. Since file uploads can be large or performed in a loop, this could lead to socket exhaustion or performance degradation.

If this becomes a performance bottleneck, consider creating a secondary `requests.Session` (or utilizing a custom adapter / retry configuration) to preserve connection pooling while still bypassing the automated retry adapter.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/filesystem.py [hunk 12]
```diff
@@ -82,11 +142,30 @@ def write(
...
+                data=data,
                 headers={"Content-Type": "application/octet-stream"},
                 timeout=timeout,
+                _disable_retries=disable_retries,
+            )
+        elif not isinstance(content, bytes):
+            prefix, suffix, content_type = _multipart_framing(path)
+            # POST is intentionally absent from RETRYABLE_METHODS. Making it
+            # retryable would replay this consumed generator as an empty body.
+            self.connector.send_request(
```
Comment (barney-s) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/filesystem.py:163]: For standard legacy multipart POST uploads, `_disable_retries=True` is not currently passed to `send_request()`. 

Although `POST` is not in `RETRYABLE_METHODS`, `requests`' default retry adapter (via `urllib3`) can still retry connection-level failures (like DNS failures, connection resets, etc.) by default. If a connection failure occurs after the body generator starts yielding, retrying would send an empty or corrupted body because the stream is partially or fully consumed.

To guarantee absolute safety, consider passing `_disable_retries=True` here as well, which will use `requests.request` without any retry adapters.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/async_filesystem.py [hunk 13]
```diff
@@ -60,17 +89,38 @@ async def write(
...
-                headers={"Content-Type": "application/octet-stream"},
-                timeout=timeout,
+                "POST", "upload", files=files_payload, timeout=timeout
             )
         else:
-            files_payload = {"file": (path, content)}
+            prefix, suffix, content_type = _multipart_framing(path)
+            # This async iterator is one-shot. AsyncSandboxConnector must not
+            # add automatic POST retries, which would replay it as an empty body.
             await self.connector.send_request(
```
Comment (barney-s) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/async_filesystem.py:118]: Consider passing `_disable_retries=True` here to remain consistent with the synchronous `Filesystem` and sandboxd PUT stream writes. Explicitly disabling retries for the multipart POST stream ensures that even if retry behaviors or retry methods are adjusted in the connector in the future, streaming uploads are guaranteed to remain one-shot to prevent sending corrupted/consumed request bodies.

**Review by barney-s — COMMENTED** [review]
Both synchronous and asynchronous paths are covered, and existing bytes/str behavior is preserved. Test coverage is good.

Few technical recommendations regarding retry safety and conn pooling

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/async_connector.py [hunk 14]
```diff
@@ -274,15 +275,16 @@ async def send_request(
                     headers["X-Sandbox-Pod-IP"] = self._pod_ip
 
         last_response: httpx.Response | None = None
-        for attempt in range(MAX_RETRIES + 1):
+        max_retries = 0 if disable_retries else MAX_RETRIES
```
Comment (janetkuo) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/async_connector.py]: When `disable_retries=True`, `max_retries` is 0 and only 1 attempt runs. However, lines 334–336 still reference `MAX_RETRIES + 1`:

```python
logger.error(f"All {MAX_RETRIES + 1} attempts failed for {url}")
raise SandboxRequestError(
    f"Failed to communicate with the sandbox at {url} after {MAX_RETRIES + 1} attempts.",
...
```

For one-shot requests that fail (e.g. 5xx or connection failures), this logs and raises `"after 4 attempts"` despite only running once. Could we compute `total_attempts = max_retries + 1` (or format based on `max_retries`) so the error message accurately reflects the single attempt?

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/async_filesystem.py [hunk 15]
```diff
@@ -12,19 +12,44 @@
...
 from k8s_agent_sandbox.trace_manager import async_trace_span, trace
 
 
+async def _iter_multipart_body(
+    prefix: bytes,
+    content: BinaryIO,
+    suffix: bytes,
+) -> AsyncIterator[bytes]:
+    yield prefix
+    while chunk := await asyncio.to_thread(_read_binary_chunk, content):
```
Comment (janetkuo) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/async_filesystem.py]: We can reuse `_iter_binary_chunks(content)` here:

```python
async def _iter_multipart_body(
    prefix: bytes,
    content: BinaryIO,
    suffix: bytes,
) -> AsyncIterator[bytes]:
    yield prefix
    async for chunk in _iter_binary_chunks(content):
        yield chunk
    yield suffix
```

This keeps it DRY and mirrors the sync implementation's `yield from _iter_binary_chunks(content)`.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/async_filesystem.py (hunk 10 above)
Comment (janetkuo) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/async_filesystem.py]: `content` accepts `BinaryIO`, which implies synchronous readers like `open(..., "rb")` or `BytesIO`. However, users working in `AsyncFilesystem` may attempt to pass an async file-like object (such as `aiofiles` or `anyio` file streams) because they have a `.read()` method. When passed here, `content.read()` returns a coroutine rather than bytes, failing downstream in `_read_binary_chunk` with a `TypeError`.

We should clarify in the docstring and type hints that only synchronous binary file streams are supported here, or note if async stream support is planned separately.

**Comment by aditya-shantanu** [comment]
/lgtm

Re-reviewed at 58fc756 — delta since a1c91c8 addresses barney-s's notes: streamed uploads now reuse a pooled no-retry session instead of bypassing the connection pool, and the async multipart stream passes `_disable_retries=True` to match sync.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/filesystem.py [hunk 16]
```diff
@@ -12,11 +12,16 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import io
```
Comment (janetkuo) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/filesystem.py]: Nit: `import io` is placed before the module docstring on line 16. Standard convention in the repo is placing the module docstring right after the Apache-2.0 license header, followed by imports.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/async_filesystem.py [hunk 17]
```diff
@@ -12,18 +12,29 @@
...
 
 from k8s_agent_sandbox.async_connector import AsyncSandboxConnector
+from k8s_agent_sandbox.files.filesystem import (
+    Filesystem,
+    _multipart_framing,
+    _read_binary_chunk,
+    _validate_binary_stream,
+)
 from k8s_agent_sandbox.exceptions import SandboxRequestError
 from k8s_agent_sandbox.files.filesystem import (
```
Comment (janetkuo) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/async_filesystem.py]: `Filesystem`, `_multipart_framing`, `_read_binary_chunk`, and `_validate_binary_stream` are imported twice from `k8s_agent_sandbox.files.filesystem` across lines 24–38. We can consolidate them into a single import block.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/async_filesystem.py [hunk 18]
```diff
@@ -50,6 +61,23 @@ async def _write_all(destination: AsyncBinaryWriter, content: bytes) -> int:
     return written
 
 
+async def _iter_multipart_body(
+    prefix: bytes,
+    content: BinaryIO,
+    suffix: bytes,
+) -> AsyncIterator[bytes]:
+    yield prefix
+    while chunk := await asyncio.to_thread(_read_binary_chunk, content):
```
Comment (janetkuo) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/async_filesystem.py]: We can reuse `_iter_binary_chunks(content)` here:

```python
async def _iter_multipart_body(
    prefix: bytes,
    content: BinaryIO,
    suffix: bytes,
) -> AsyncIterator[bytes]:
    yield prefix
    async for chunk in _iter_binary_chunks(content):
        yield chunk
    yield suffix
```

This keeps it DRY and mirrors the sync implementation's `yield from _iter_binary_chunks(content)`.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/async_filesystem.py [hunk 19]
```diff
@@ -63,18 +91,22 @@ def __init__(
     @async_trace_span("write")
     async def write(
         self,
-        path: str, content: bytes | str,
+        path: str,
+        content: bytes | str | BinaryIO,
```
Comment (janetkuo) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/async_filesystem.py]: `content` accepts `BinaryIO`, which implies synchronous readers like `open(..., "rb")` or `BytesIO`. However, users working in `AsyncFilesystem` may attempt to pass an async file-like object (such as `aiofiles` or `anyio` file streams) because they have a `.read()` method. When passed here, `content.read()` returns a coroutine rather than bytes, failing downstream in `_read_binary_chunk` with a `TypeError`.

We should clarify in the docstring that synchronous binary file streams are expected here.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/async_connector.py [hunk 20]
```diff
@@ -306,7 +307,8 @@ async def send_request(
             else httpx.USE_CLIENT_DEFAULT
         )
         last_response: httpx.Response | None = None
-        for attempt in range(MAX_RETRIES + 1):
+        max_retries = 0 if disable_retries else MAX_RETRIES
```
Comment (janetkuo) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/async_connector.py]: When `disable_retries=True`, `max_retries` is 0 and only 1 attempt runs. Note that later in this function (lines 385–387), the failure logs and exception message still hardcode `MAX_RETRIES + 1`:

```python
logger.error(f"All {MAX_RETRIES + 1} attempts failed for {url}")
raise SandboxRequestError(
    f"Failed to communicate with the sandbox at {url} after {MAX_RETRIES + 1} attempts.",
...
```

For one-shot requests that fail, this logs and raises `"after 6 attempts"` despite only running once. We could compute `total_attempts = max_retries + 1` so the error message accurately reflects the single attempt.

**Review by aditya-shantanu — COMMENTED** [review]
Re-reviewed at 69c2fab. Range-diff against 58fc756 confirms the push is a pure rebase over #1643 (streaming downloads) — no logic changes — except one conflict-resolution artifact noted inline. Will re-tag once that's fixed.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/filesystem.py (hunk 16 above)
Comment (aditya-shantanu) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/filesystem.py]: Rebase artifact: `import io` landed above the module docstring, so the string on the next line is no longer `__doc__`. Please move the docstring back above the imports.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/async_filesystem.py [hunk 21]
```diff
@@ -12,18 +12,29 @@
...
+)
 from k8s_agent_sandbox.exceptions import SandboxRequestError
 from k8s_agent_sandbox.files.filesystem import (
     Filesystem,
+    _multipart_framing,
+    _read_binary_chunk,
     _STREAM_CHUNK_SIZE,
     _sandboxd_files_endpoint,
+    _validate_binary_stream,
 )
```
Comment (janetkuo) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/async_filesystem.py]: `Filesystem`, `_multipart_framing`, `_read_binary_chunk`, and `_validate_binary_stream` are imported twice from `k8s_agent_sandbox.files.filesystem` across lines 24–38. We can consolidate them into a single import block.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/async_filesystem.py (hunk 18 above)
Comment (janetkuo) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/async_filesystem.py]: We can reuse `_iter_binary_chunks(content)` here:

```python
async def _iter_multipart_body(
    prefix: bytes,
    content: BinaryIO,
    suffix: bytes,
) -> AsyncIterator[bytes]:
    yield prefix
    async for chunk in _iter_binary_chunks(content):
        yield chunk
    yield suffix
```

This keeps it DRY and mirrors the sync implementation's `yield from _iter_binary_chunks(content)`.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/async_filesystem.py (hunk 19 above)
Comment (janetkuo) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/files/async_filesystem.py]: `content` accepts `BinaryIO`, which implies synchronous readers like `open(..., "rb")` or `BytesIO`. However, users working in `AsyncFilesystem` may attempt to pass an async file-like object (such as `aiofiles` or `anyio` file streams) because they have a `.read()` method. When passed here, `content.read()` returns a coroutine rather than bytes, failing downstream in `_read_binary_chunk` with a `TypeError`.

We should clarify in the docstring that synchronous binary file streams are expected here.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/async_connector.py (hunk 20 above)
Comment (janetkuo) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/async_connector.py]: When `disable_retries=True`, `max_retries` is 0 and only 1 attempt runs. Note that later in this function (lines 385–387), the failure logs and exception message still hardcode `MAX_RETRIES + 1`:

```python
logger.error(f"All {MAX_RETRIES + 1} attempts failed for {url}")
raise SandboxRequestError(
    f"Failed to communicate with the sandbox at {url} after {MAX_RETRIES + 1} attempts.",
...
```

For one-shot requests that fail, this logs and raises `"after 6 attempts"` despite only running once. We could compute `total_attempts = max_retries + 1` so the error message accurately reflects the single attempt.
